src/main/webapp/assets/script.js

```diff
@@ -72,12 +73,89 @@ function onLoad() {
   collegeLocation = fetchFakeCollegeLocation();
 
   // Add the embedded map to the page.
-  const script = document.createElement('script');
-  script.src = `https://maps.googleapis.com/maps/api/js?key=AIzaSyB_iMemNeUtO8D7kGTmagTgluRlsiFTcDA&callback=initMap`;
-  script.defer = true;
-  script.async = true;
-  window.initMap = initMap;
-  document.head.appendChild(script);
+  getSecretFor('javascript-maps-api').then((key) => {
+    const script = document.createElement('script');
+    script.src = `https://maps.googleapis.com/maps/api/js?key=${key}&callback=initMap`;
+    script.defer = true;
+    script.async = true;
+    window.initMap = initMap;
+    document.head.appendChild(script);
+  });
+}
+
+/**
+ * Gets the secret value corresponding to a secret ID from GCP secrets store.
+ * @param {string} secretid - The secret's id, as defined in the secrets store.
+ */
+async function getSecretFor(secretid) {
+  try {
+    const response = await fetch('/secret?id=' + secretid, {method: 'POST'});
+    if (!response.ok) {
+      throw new Error(response.status);
+    } else {
+      return await response.text();
+    }
+  } catch (err) {
+    console.warn(err);
+    return;
+  }
+}
+
+/**
+ * Translates a location from its name to a pair of latitude and longitudes.
+ * @param {string} address - The address to translate to lat/long.
+ * @return {LocationInfo} or null if no such location exists or an error occurs.
+ */
+async function translateLocationToLatLong(address) {
+  try {
+    const response = await fetch('/translateLocation', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
+      },
+      body: createSearchParamsFromObject({location: address}),
+    });
+
+    // If we get a non-200 status response, then fail.
+    if (!response.ok) {
+      const error = await response.text();
+      throw new Error(response.status + ' ' + error);
+    }
+
+    const result = await response.json();
+    console.log(result);
+
+    // If our result is empty, then we didn't have any results at all.
+    if (Object.keys(result).length === 0) {
+      throw new Error('No results returned with query');
+    }
+
+    // If our result doesn't have certain properties, then there's no valid lat/long.
+    if (!('geometry' in result) || !('location' in result.geometry)) {
+      throw new Error('No latitude/longitude was returned with that query.');
```

Comment:
I noticed in a lot of the catching exceptions, the usual behavior is to throw/set to a 400 or a new error in the console. In the future, will you be adding a page to update users that their location is invalid so the user doesn't get confused with that happened?

Reply:
This is just a wrapper for `TranslateLocationServlet.java`, so the behavior of what the app should do when the user enters an invalid address is up to the caller of this function. The function returns null if the provided address is invalid, so anybody who calls this function should check its result for null and then take appropriate action, such as displaying an error to the user and asking them to be more specific. I don't want to hardcode any page-specific behavior into this function, because it should be reusable from anywhere within the app.